The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9700004801/job/26770388853) [(plain)](https://github.com/rust-lang/rust/commit/c9870cfa4bed1163be4e9fde814d79860f3caa57/checks/26770388853/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
error[E0034]: multiple applicable items in scope
    --> compiler/rustc_middle/src/ty/context.rs:545:14
     |
545  |         self.supertrait_def_ids(trait_def_id)
     |              ^^^^^^^^^^^^^^^^^^ multiple `supertrait_def_ids` found
note: candidate #1 is defined in an impl for the type `context::TyCtxt<'tcx>`
    --> compiler/rustc_middle/src/ty/context.rs:2470:5
     |
     |
2470 |     fn supertrait_def_ids(self, trait_def_id: DefId) -> impl Iterator<Item = DefId> + 'tcx {
note: candidate #2 is defined in an impl for the type `context::TyCtxt<'tcx>`
    --> compiler/rustc_middle/src/ty/vtable.rs:46:5
     |
     |
46   |     pub fn supertrait_def_ids(self, trait_def_id: DefId) -> SupertraitDefIds<'tcx> {

   Compiling rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
   Compiling rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
error[E0592]: duplicate definitions with name `supertrait_def_ids`
error[E0592]: duplicate definitions with name `supertrait_def_ids`
    --> compiler/rustc_middle/src/ty/vtable.rs:46:5
     |
46   |     pub fn supertrait_def_ids(self, trait_def_id: DefId) -> SupertraitDefIds<'tcx> {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `supertrait_def_ids`
    ::: compiler/rustc_middle/src/ty/context.rs:2470:5
     |
     |
2470 |     fn supertrait_def_ids(self, trait_def_id: DefId) -> impl Iterator<Item = DefId> + 'tcx {
     |     -------------------------------------------------------------------------------------- other definition for `supertrait_def_ids`
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0034]: multiple applicable items in scope
    --> compiler/rustc_middle/src/ty/context.rs:2446:14
     |
     |
2446 |         self.supertrait_def_ids(trait_def_id).any(|trait_did| {
     |              ^^^^^^^^^^^^^^^^^^ multiple `supertrait_def_ids` found
note: candidate #1 is defined in an impl for the type `context::TyCtxt<'tcx>`
    --> compiler/rustc_middle/src/ty/context.rs:2470:5
     |
     |
2470 |     fn supertrait_def_ids(self, trait_def_id: DefId) -> impl Iterator<Item = DefId> + 'tcx {
note: candidate #2 is defined in an impl for the type `context::TyCtxt<'tcx>`
    --> compiler/rustc_middle/src/ty/vtable.rs:46:5
     |
     |
46   |     pub fn supertrait_def_ids(self, trait_def_id: DefId) -> SupertraitDefIds<'tcx> {

Some errors have detailed explanations: E0034, E0592.
For more information about an error, try `rustc --explain E0034`.
error: could not compile `rustc_middle` (lib) due to 3 previous errors
```

</details>

--- a/compiler/rustc_middle/src/ty/context.rs
+++ b/compiler/rustc_middle/src/ty/context.rs
@@ -2466,8 +2466,9 @@ impl<'tcx> TyCtxt<'tcx> {
 
     /// Computes the def-ids of the transitive supertraits of `trait_def_id`. This (intentionally)
     /// does not compute the full elaborated super-predicates but just the set of def-ids. It is used
-    /// to identify which traits may define a given associated type to help avoid cycle errors.
-    fn supertrait_def_ids(self, trait_def_id: DefId) -> impl Iterator<Item = DefId> + 'tcx {
+    /// to identify which traits may define a given associated type to help avoid cycle errors,
+    /// and to make size estimates for vtable layout computation.
+    pub fn supertrait_def_ids(self, trait_def_id: DefId) -> impl Iterator<Item = DefId> + 'tcx {
         let mut set = FxHashSet::default();
         let mut stack = vec![trait_def_id];
 

--- a/compiler/rustc_middle/src/ty/vtable.rs
+++ b/compiler/rustc_middle/src/ty/vtable.rs
@@ -3,8 +3,6 @@ use std::fmt;
 use crate::mir::interpret::{alloc_range, AllocId, Allocation, Pointer, Scalar};
 use crate::ty::{self, Instance, PolyTraitRef, Ty, TyCtxt};
 use rustc_ast::Mutability;
-use rustc_data_structures::fx::FxHashSet;
-use rustc_hir::def_id::DefId;
 use rustc_macros::HashStable;
 
 #[derive(Clone, Copy, PartialEq, HashStable)]
@@ -42,44 +40,11 @@ impl<'tcx> fmt::Debug for VtblEntry<'tcx> {
 impl<'tcx> TyCtxt<'tcx> {
     pub const COMMON_VTABLE_ENTRIES: &'tcx [VtblEntry<'tcx>] =
         &[VtblEntry::MetadataDropInPlace, VtblEntry::MetadataSize, VtblEntry::MetadataAlign];
-
-    pub fn supertrait_def_ids(self, trait_def_id: DefId) -> SupertraitDefIds<'tcx> {
-        SupertraitDefIds {
-            tcx: self,
-            stack: vec![trait_def_id],
-            visited: Some(trait_def_id).into_iter().collect(),
-        }
-    }
 }
 
 pub const COMMON_VTABLE_ENTRIES_DROPINPLACE: usize = 0;
 pub const COMMON_VTABLE_ENTRIES_SIZE: usize = 1;
 pub const COMMON_VTABLE_ENTRIES_ALIGN: usize = 2;
-
-pub struct SupertraitDefIds<'tcx> {
-    tcx: TyCtxt<'tcx>,
-    stack: Vec<DefId>,
-    visited: FxHashSet<DefId>,
-}
-
-impl Iterator for SupertraitDefIds<'_> {
-    type Item = DefId;
-
-    fn next(&mut self) -> Option<DefId> {
-        let def_id = self.stack.pop()?;
-        let predicates = self.tcx.explicit_super_predicates_of(def_id);
-        let visited = &mut self.visited;
-        self.stack.extend(
-            predicates
-                .predicates
-                .iter()
-                .filter_map(|(pred, _)| pred.as_trait_clause())
-                .map(|trait_ref| trait_ref.def_id())
-                .filter(|&super_def_id| visited.insert(super_def_id)),
-        );
-        Some(def_id)
-    }
-}
 
 // Note that we don't have access to a self type here, this has to be purely based on the trait (and
 // supertrait) definitions. That means we can't call into the same vtable_entries code since that

